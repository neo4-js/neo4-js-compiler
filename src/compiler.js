// @flow

import { parse } from "./parser.js";
import prettier from "prettier";

/**
 * Validate if default values match the scalar type
 * Validate if used relations are defined
 */
function validateModel(model, relations) {
  model.properties.forEach(p => {
    // Validate if default values match the scalar type
    if (p.scalar && p.defaultValue && p.type !== p.defaultValue.type) {
      throw new Error(`Default value doesn't match type of property.

${model.name}.${p.name} is of type ${p.type} and your default value ${p
        .defaultValue.value} is of type ${p.defaultValue.type}.`);
    }

    // Validate if used relations are defined
    if (p.type === "Relation" && !relations.find(r => r.name === p.relation)) {
      throw new Error(
        `Relation ${p.relation} on ${model.name}.${p.name} not defined`
      );
    }
  });
}

/**
 * Validate if model is valid
 * Validate if model name is unique
 */
function validateModels(models, relations) {
  // Validate if model names are unique
  models.forEach(m => {
    if (models.filter(b => b.name === m.name).length !== 1)
      throw new Error(
        `Model names must be unqiue. ${m.name} is defined multiple times`
      );
  });

  models.forEach(m => validateModel(m, relations));
}

function printRelationHeader(r) {
  return `relation ${r.name} from ${r.from} to ${r.to} via "${r.via}"`;
}

// In a week I'll have no idea what I did here
function validateRelation(r, models) {
  // Check if connections use from/to models
  const usedModels = models.filter(m =>
    m.properties.find(p => p.type === "Relation" && p.relation === r.name)
  );
  const checkModel = m => {
    if (m.name !== r.from && m.name !== r.to) {
      throw new Error(
        `Relation ${r.name} can't be used within ${m.name} model because it's neither used as from or to model

${printRelationHeader(r)} {`
      );
    }
    const prop = m.properties.find(
      p => p.type === "Relation" && p.relation === r.name
    );

    /**
     * Check if the destination relations of a model property is either defined in relation.to or relation.from
     */
    if (prop.to !== r.from && prop.to !== r.to) {
      throw new Error(
        `${m.name}.${prop.name} needs to point ${r.from !== r.to
          ? `either to ${r.from} or ${r.to}`
          : `to ${r.from}`} as defined in

${printRelationHeader(r)} {

Change
${prop.name}: has ${prop.many ? "many" : "one"} ${prop.to} using ${r.name}
${r.from !== r.to
          ? `either to
${prop.name}: has ${prop.many ? "many" : "one"} ${r.from} using ${r.name}
or
${prop.name}: has ${prop.many ? "many" : "one"} ${r.to} using ${r.name}`
          : `to
${prop.name}: has ${prop.many ? "many" : "one"} ${r.from} using ${r.name}`}`
      );
    }
  };
  usedModels.forEach(checkModel);

  // Check if used models are defined
  const checkDefined = dir => {
    if (!models.find(m => m.name === r[dir])) {
      throw new Error(
        `Model ${r[dir]} not defined

${printRelationHeader(r)}`
      );
    }
  };
  checkDefined("to");
  checkDefined("from");
}

function validateRelations(models, relations) {
  // Validate if relation names are unique
  relations.forEach(a => {
    if (relations.filter(b => a.name === b.name).length !== 1)
      throw new Error(
        `Relation names must be unqiue. ${a.name} is defined multiple times`
      );
  });

  relations.forEach(r => validateRelation(r, models));
}

function compileRelation(r) {
  return [
    `const ${r.name} = relation`,
    `  .from(() => ${r.from})`,
    `  .to(() => ${r.to})`,
    `  .via("${r.via}");`,
  ].join("\n");
}

function compileRelationProperty(model, prop, relations) {
  return [
    `@${prop.many
      ? "hasMany"
      : "hasOne"}(() => ${prop.to}, ${prop.relation}${prop.direction
      ? `, "${prop.direction}"`
      : ""})`,
    `// prettier-ignore`,
    `${prop.name}/*: Has${prop.many
      ? "Many"
      : "One"}Actions<${prop.to}Props, ${prop.to}Instance> */;`,
  ].join("\n");
}

function compileProperty(prop) {
  let type = prop.type.toLowerCase();
  return `  ${prop.name}: ${type},`;
}

function compileModel(model, relations) {
  const template = [
    "/*::",
    `export type ${model.name}Props = {|`,
    model.properties
      .filter(p => p.type !== "Relation")
      .map(compileProperty)
      .join("\n"),
    "|}",
    "*/",
    "",
    `class ${model.name}Model extends Model/*:: <${model.name}Props, ${model.name}Instance>*/ { }`,
    `export const ${model.name}/*: ${model.name}Model*/ = new ${model.name}Model("${model.name}");`,
    `@model(${model.name})`,
    `export class ${model.name}Instance extends ModelInstance/*:: <${model.name}Props>*/ {`,
    model.properties
      .filter(p => p.type === "Relation")
      .map(r => compileRelationProperty(model, r, relations))
      .join("\n"),
    "}",
  ].join("\n");

  return template;
}

export function compile(str: string) {
  const schema = parse(str);

  const models = schema.filter(m => m.type === "Model");
  const relations = schema.filter(r => r.type === "Relation");
  validateModels(models, relations);
  validateRelations(models, relations);
  const compiledRelations = relations.map(compileRelation).join("\n");
  const compiledModels = models.map(m => compileModel(m, relations)).join("\n");
  const compiled = [
    "/**",
    " * @flow",
    " */",
    `import { Model, ModelInstance, model, relation, hasMany, hasOne }Â from "neo4-js";`,
    `import type { StringProperty, NumberProperty, HasManyActions, HasOneActions } from "neo4-js";`,
    "",
    compiledRelations,
    "",
    compiledModels,
  ].join("\n");

  return prettier.format(compiled, { trailingComma: "es5" });
}
