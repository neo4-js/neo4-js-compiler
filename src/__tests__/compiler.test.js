import { compile } from "../compiler.js";

describe("compiler", () => {
  // TODO: check other types than string
  it("should throw an error if default value doesn't match type", () => {
    const schema = `
      type User {
        name: String = 1
      }
    `;

    try {
      const result = compile(schema);
      expect(true).toEqual(false);
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });

  it("should throw an error if model names are not unique", () => {
    const schema = `
      type User {
        name: String
      }

      type User {
        email: String
      }
    `;

    try {
      const result = compile(schema);
      expect(true).toEqual(false);
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });

  it("should throw an error if relation is not defined but used", () => {
    let schema = `
      type User {
        name: String
        createdTasks: has many Task using UserCreatedTasks
      }

      type Task {
        title: String
        creator: has one User using UserCreatedTasks
      }
    `;

    try {
      const result = compile(schema);
      expect(true).toEqual(false);
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });

  it("should throw an error if wrong destination model is used within a property definition", () => {
    let schema = `
      type User {
        name: String
        tasks: has many Task using UserTasksRelation
      }

      type Task {
        title: String
        creator: has one Comment using UserTasksRelation
      }

      type Comment { }

      relation UserTasksRelation from User to Task via "created" { }
    `;

    try {
      const result = compile(schema);
      expect(true).toEqual(false);
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });

  it("should throw an error if a relation is used in a model which is not equal to from/to model names from relation", () => {
    let schema = `
      type User {
        name: String
      }

      type Task {
        title: String
        creator: has one User using UserTasksRelation
      }

      type Comment {
        tasks: has many Task using UserTasksRelation
      }

      relation UserTasksRelation from User to Task via "created" { }
    `;

    try {
      const result = compile(schema);
      expect(true).toEqual(false);
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });

  it("should throw an error if wrong destination model is used within a property definition in a recursive relation", () => {
    let schema = `
      type User {
        name: String
        tasks: has many Task using UserUserRelation
      }

      relation UserUserRelation from User to User via "created" { }
    `;

    try {
      const result = compile(schema);
      expect(true).toEqual(false);
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });

  it("should compile simplest possible schema", () => {
    let schema = `type User { name: String }`;
    try {
      const result = compile(schema);
      expect(result).toMatchSnapshot();
    } catch (err) {
      expect(err).toBeFalsy();
    }
  });

  it("should compile complex schemas", () => {
    let schema = `
      type User {
        name: String,
        age: Number,
        activated: Boolean,
      }

      type Task {
        title: String = "empty",
        done: Boolean = false,
      }
    `;
    try {
      const result = compile(schema);
      expect(result).toMatchSnapshot();
    } catch (err) {
      expect(err).toBeFalsy();
    }
  });

  it("should compile schema with simple relation", () => {
    let schema = `
      type User {
        name: String,
        age: Number,
        activated: Boolean,
        createdTasks: has many Task using TaskCreatorRelation,
      }

      type Task {
        title: String = "empty",
        done: Boolean = false,
        creator: has one User using TaskCreatorRelation
      }

      relation TaskCreatorRelation from User to Task via "created" { }
    `;
    try {
      const result = compile(schema);
      expect(result).toMatchSnapshot();
    } catch (err) {
      expect(err).toBeFalsy();
    }
  });

  it("should compile schema with recursive relation", () => {
    let schema = `
      type User {
        name: String,
        age: Number,
        activated: Boolean,
        parent: has one User using ParentRelation out,
        children: has many User using ParentRelation in,
      }

      relation ParentRelation from User to User via "parent" { }
    `;
    try {
      const result = compile(schema);
      expect(result).toMatchSnapshot();
    } catch (err) {
      expect(err).toBeFalsy();
    }
  });
});
