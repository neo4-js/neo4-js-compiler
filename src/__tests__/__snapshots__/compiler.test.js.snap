// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler should compile complex schemas 1`] = `
"/**
 * @flow
 */
import {
  Model,
  ModelInstance,
  model,
  relation,
  hasMany,
  hasOne,
} from \\"neo4-js\\";
import type {
  StringProperty,
  NumberProperty,
  HasManyActions,
  HasOneActions,
} from \\"neo4-js\\";

/*::
export type UserProps = {|
  name: string,
  age: number,
  activated: boolean,
|}
*/

class UserModel extends Model /*:: <UserProps, UserInstance>*/ {
}
export const User /*: UserModel*/ = new UserModel(\\"User\\");
@model(User)
export class UserInstance extends ModelInstance /*:: <UserProps>*/ {
}
/*::
export type TaskProps = {|
  title: string,
  done: boolean,
|}
*/

class TaskModel extends Model /*:: <TaskProps, TaskInstance>*/ {
}
export const Task /*: TaskModel*/ = new TaskModel(\\"Task\\");
@model(Task)
export class TaskInstance extends ModelInstance /*:: <TaskProps>*/ {
}
"
`;

exports[`compiler should compile schema with recursive relation 1`] = `
"/**
 * @flow
 */
import {
  Model,
  ModelInstance,
  model,
  relation,
  hasMany,
  hasOne,
} from \\"neo4-js\\";
import type {
  StringProperty,
  NumberProperty,
  HasManyActions,
  HasOneActions,
} from \\"neo4-js\\";

const ParentRelation = relation
  .from(() => User)
  .to(() => User)
  .via(\\"parent\\");

/*::
export type UserProps = {|
  name: string,
  age: number,
  activated: boolean,
|}
*/

class UserModel extends Model /*:: <UserProps, UserInstance>*/ {
}
export const User /*: UserModel*/ = new UserModel(\\"User\\");
@model(User)
export class UserInstance extends ModelInstance /*:: <UserProps>*/ {
  @hasOne(() => User, ParentRelation, \\"out\\")
  // prettier-ignore
  parent /*: HasOneActions<UserProps, UserInstance> */;
  @hasMany(() => User, ParentRelation, \\"in\\")
  // prettier-ignore
  children /*: HasManyActions<UserProps, UserInstance> */;
}
"
`;

exports[`compiler should compile schema with simple relation 1`] = `
"/**
 * @flow
 */
import {
  Model,
  ModelInstance,
  model,
  relation,
  hasMany,
  hasOne,
} from \\"neo4-js\\";
import type {
  StringProperty,
  NumberProperty,
  HasManyActions,
  HasOneActions,
} from \\"neo4-js\\";

const TaskCreatorRelation = relation
  .from(() => User)
  .to(() => Task)
  .via(\\"created\\");

/*::
export type UserProps = {|
  name: string,
  age: number,
  activated: boolean,
|}
*/

class UserModel extends Model /*:: <UserProps, UserInstance>*/ {
}
export const User /*: UserModel*/ = new UserModel(\\"User\\");
@model(User)
export class UserInstance extends ModelInstance /*:: <UserProps>*/ {
  @hasMany(() => Task, TaskCreatorRelation)
  // prettier-ignore
  createdTasks /*: HasManyActions<TaskProps, TaskInstance> */;
}
/*::
export type TaskProps = {|
  title: string,
  done: boolean,
|}
*/

class TaskModel extends Model /*:: <TaskProps, TaskInstance>*/ {
}
export const Task /*: TaskModel*/ = new TaskModel(\\"Task\\");
@model(Task)
export class TaskInstance extends ModelInstance /*:: <TaskProps>*/ {
  @hasOne(() => User, TaskCreatorRelation)
  // prettier-ignore
  creator /*: HasOneActions<UserProps, UserInstance> */;
}
"
`;

exports[`compiler should compile simplest possible schema 1`] = `
"/**
 * @flow
 */
import {
  Model,
  ModelInstance,
  model,
  relation,
  hasMany,
  hasOne,
} from \\"neo4-js\\";
import type {
  StringProperty,
  NumberProperty,
  HasManyActions,
  HasOneActions,
} from \\"neo4-js\\";

/*::
export type UserProps = {|
  name: string,
|}
*/

class UserModel extends Model /*:: <UserProps, UserInstance>*/ {
}
export const User /*: UserModel*/ = new UserModel(\\"User\\");
@model(User)
export class UserInstance extends ModelInstance /*:: <UserProps>*/ {
}
"
`;

exports[`compiler should throw an error if a relation is used in a model which is not equal to from/to model names from relation 1`] = `
[Error: Relation UserTasksRelation can't be used within Comment model because it's neither used as from or to model

relation UserTasksRelation from User to Task via "created" {]
`;

exports[`compiler should throw an error if default value doesn't match type 1`] = `
[Error: Default value doesn't match type of property.

User.name is of type String and your default value 1 is of type Number.]
`;

exports[`compiler should throw an error if model names are not unique 1`] = `[Error: Model names must be unqiue. User is defined multiple times]`;

exports[`compiler should throw an error if relation is not defined but used 1`] = `[Error: Relation UserCreatedTasks on User.createdTasks not defined]`;

exports[`compiler should throw an error if wrong destination model is used within a property definition 1`] = `
[Error: Task.creator needs to point either to User or Task as defined in

relation UserTasksRelation from User to Task via "created" {

Change
creator: has one Comment using UserTasksRelation
either to
creator: has one User using UserTasksRelation
or
creator: has one Task using UserTasksRelation]
`;

exports[`compiler should throw an error if wrong destination model is used within a property definition in a recursive relation 1`] = `
[Error: User.tasks needs to point to User as defined in

relation UserUserRelation from User to User via "created" {

Change
tasks: has many Task using UserUserRelation
to
tasks: has many User using UserUserRelation]
`;
